File: Data Structures & Algorithms/search-for-word-ii/submission-3.java
class Trie {
    Trie[] child;
    boolean end;

    public Trie() {
        child = new Trie[26];
        end = false;
    }

    private void add (String word) {
        Trie curr = this;
        for (char c : word.toCharArray()) {
            int idx = c - 'a';
            if (curr.child[idx] == null) {
                curr.child[idx] = new Trie();
            }
            curr = curr.child[idx];
        }
        curr.end = true;
    }
}

class Solution {
    int ROWS, COLS;
    boolean[][] visited;
    Set<String> result;

    public List<String> findWords(char[][] board, String[] words) {
        Trie root = new Trie();
        for (String word : words) {
            root.add (word);
        }

        ROWS = board.length;
        COLS = board[0].length;
        visited = new boolean[ROWS][COLS];
        result = new HashSet<>();

        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                dfs (board, r, c, root, "");
            }
        }
        return new ArrayList<>(result);
    }

    private void dfs (char[][] board, int r, int c, Trie  node, String word) {
        if (r < 0 || r >= ROWS || c < 0 || c >= COLS || visited[r][c]) {
            return;
        }

        char curr = board[r][c];        
        int idx = curr - 'a';
        node = node.child[idx];
        if (node == null) {
            return;
        }
        word += curr;
        visited[r][c] = true;
        
        if (node.end) {
            result.add(word);
        }

        dfs (board, r + 1, c, node, word);
        dfs (board, r - 1, c, node, word);
        dfs (board, r, c + 1, node, word);
        dfs (board, r, c - 1, node, word);

        visited[r][c] = false;
    }
}

File: Data Structures & Algorithms/count-number-of-islands/submission-3.cpp
class Solution {
private:
    int dirs[4][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    int ROWS, COLS;

public:
    int numIslands(vector<vector<char>>& grid) {
        ROWS = grid.size();
        COLS = grid[0].size();

        int islands = 0;
        for(int row = 0; row < ROWS; row++) {
            for(int col = 0; col < COLS; col++) {
                if(grid[row][col] == '1') {
                    islands++;
                    dfs(grid, row, col);
                }
            }
        }

        return islands;
    }

    void dfs(vector<vector<char>>& grid, int row, int col) {
        if(row < 0 || row >= ROWS || col < 0 || col >= COLS || grid[row][col] == '0') {
            return;
        }

        grid[row][col] = '0';
        for(int i = 0; i < 4; i++) {
            dfs(grid, row + dirs[i][0], col + dirs[i][1]);
        }
    }
};
